File: go/functions/struct_example.go
package main

import "fmt"

type Person struct {
	Name string
	Age  int
}

func (p *Person) Greet() string {
	return fmt.Sprintf("Hello, I'm %s and I'm %d years old", p.Name, p.Age)
}

func (p *Person) Birthday() string {
	p.Age++
	return fmt.Sprintf("Happy birthday! Now %d years old", p.Age)
}

type Employee struct {
	Person
	JobTitle string
}

func (e *Employee) Work() string {
	return fmt.Sprintf("%s is working as a %s", e.Name, e.JobTitle)
}

func main() {
	person := Person{Name: "Alice", Age: 30}
	fmt.Println(person.Greet())
	fmt.Println(person.Birthday())
	fmt.Printf("%+v\n", person)

	employee := Employee{
		Person:   Person{Name: "Bob", Age: 25},
		JobTitle: "Software Engineer",
	}
	fmt.Println(employee.Greet())
	fmt.Println(employee.Work())
	fmt.Printf("%+v\n", employee)
}

File: rust/functions/struct_example.rs
struct Person {
    name: String,
    age: u32,
}

impl Person {
    fn new(name: String, age: u32) -> Self {
        Person { name, age }
    }

    fn greet(&self) -> String {
        format!("Hello, I'm {} and I'm {} years old", self.name, self.age)
    }

    fn birthday(&mut self) -> String {
        self.age += 1;
        format!("Happy birthday! Now {} years old", self.age)
    }
}

struct Employee {
    person: Person,
    job_title: String,
}

impl Employee {
    fn new(name: String, age: u32, job_title: String) -> Self {
        Employee {
            person: Person::new(name, age),
            job_title,
        }
    }

    fn work(&self) -> String {
        format!("{} is working as a {}", self.person.name, self.job_title)
    }
}

fn main() {
    let mut person = Person::new(String::from("Alice"), 30);
    println!("{}", person.greet());
    println!("{}", person.birthday());
    println!("Person {{ name: {}, age: {} }}", person.name, person.age);

    let employee = Employee::new(
        String::from("Bob"),
        25,
        String::from("Software Engineer"),
    );
    println!("{}", employee.person.greet());
    println!("{}", employee.work());
}
